assayist/client/query.py

```diff
@@ -44,3 +44,88 @@ def get_container_content_sources(build_id):
         'internal_urls': internal_urls,
         'upstream_urls': upstream_urls
     }
+
+
+def get_current_and_previous_versions(name, type_, version):
+    """
+    Find the current and previous source locations.
+
+    :param str name: the canonical name of the component
+    :param str type_: the canonical type of the component
+    :param str version: the canonical version of the source location
+    :return: a dictionary of all the previous source locations and the current source location
+    :rtype: dict
+    """
+    # TODO: Consider alternative names as well
+    query = """
+        MATCH (:Component {{canonical_name: '{name}', canonical_type: '{type}'}})
+            <-[:SOURCE_FOR]-(:SourceLocation {{canonical_version: '{version}'}})
+            -[:SUPERSEDES*0..]->(sl:SourceLocation)
+        RETURN sl
+    """.format(name=name, type=type_, version=version)
+    results, _ = neomodel.db.cypher_query(query)
+    rv = []
+    for result in results:
+        rv.append(dict(result[0]))
+    return rv
+
+
+def get_container_built_with_sources(source_locations):
+    """
+    Match container builds that used the input source locations to build the content in the image.
+
+    :param list source_locations: a list of source location dictionaries to match against
+    :return: a list of affected container build Koji IDs
+    :rtype: list
+    """
+    if not source_locations or not isinstance(source_locations, list):
+        raise InvalidInput('The input must be a list of source locations')
+
+    # Build the where clause to match all the source locations that were passed in
+    where_clause = ''
+    for source_location in source_locations:
+        if not source_location.get('url'):
+            raise InvalidInput('One of the input source locations did not contain a "url"')
+        if where_clause:
+            where_clause += ' OR '
+        where_clause += 'input_sl.url = \'' + source_location['url'] + '\''
+
+    query = """
+        // First get all the input source locations
+        MATCH (input_sl:SourceLocation) WHERE {where_clause}
+        // Then find all the source locations that embed the input source locations recursively.
+        // The resulting `sl` variable has all the input source locations and all the source
+        // locations that eventually embed the input source locations.
+        MATCH (input_sl)<-[:EMBEDS*0..]-(sl:SourceLocation)
```

Comment:
Should this be `<-[:UPSTREAM|:EMBEDS*0..]-`?

Reply:
I'll drop the `<` so that way the source location can have an `UPSTREAM` or `EMBEDS` relationship in either direction. This should allow upstream source locations to be input for this API.